src/passes/Heap2Local.cpp

```diff
@@ -444,10 +447,10 @@ struct Heap2LocalOptimizer {
       // look at something that another allocation reached, which would be in a
       // different call to this function and use a different queue (any overlap
       // between calls would prove non-exclusivity).
-      if (seen.count(child)) {
+      if (seen.count(parent)) {
```

Comment:
Is it not possible that
```
  GP
 /  \
P1   P2
 \  /
   C
```
C flows the same value to P1 and P2 so even though GP are visited twice by two different passes but they end up in the same value?

> what this check does is verify we don't look at something that another allocation reached, which would be in a different call to this function and use a different queue (any overlap between calls would prove non-exclusivity).

Not sure about this either... What do you mean by "another allocation"? Aren't we dealing with a single allocation in this function?

Reply:
Yes, a diamond like that is possible, but not handled by this pass yet. There is a TODO on (new) lines 634-5.

We deal with a single allocation in this function, yes, but we call it multiple times on different allocations (that's why the comment mentions "different call" - maybe not clear enough?). After we have seen one node be used exclusively by allocation A, it is not possible for it to be used exclusively by allocation B, so we stop.

Reply:
Ah, I didn't realize `seen` is not a local variable to this function but a member variable. Thanks.